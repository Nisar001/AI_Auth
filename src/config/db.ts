import 'dotenv/config';
import { DataSource } from 'typeorm';
import { User } from '../models/User';
import { OtpService } from '../models/OtpService';

export const AppDataSource = new DataSource({
  type: 'postgres',
  host: process.env.DB_HOST || 'localhost',
  port: parseInt(process.env.DB_PORT || '5432'),
  username: process.env.DB_USER || 'postgres',
  password: process.env.DB_PASS || '',
  database: process.env.DB_NAME || 'ai_auth_db',
  synchronize: true, // Set to false in production
  logging: false,
  entities: [User, OtpService],
  migrations: [],
  subscribers: [],
});
