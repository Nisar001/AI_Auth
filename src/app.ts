import 'reflect-metadata';
import express from 'express';
import cors from 'cors';
import dotenv from 'dotenv';
import { AppDataSource } from './config/db';
import { errorHandler, notFoundHandler } from './middlewares/errorHandler';
import { generalLimiter } from './middlewares/rateLimiter';
import authRoutes from './modules/routes/authRoutes';
import logger from './utils/logger';

// Load environment variables
dotenv.config();

const app = express();

// Security and parsing middleware
app.use(cors({
  origin: process.env.CORS_ORIGINS?.split(',') || ['http://localhost:3000'],
  credentials: true,
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'],
  allowedHeaders: ['Content-Type', 'Authorization']
}));

app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));

// Apply rate limiting to all requests
app.use(generalLimiter);

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({
    status: 'OK',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    environment: process.env.NODE_ENV || 'development'
  });
});

// API routes
app.use('/api/auth', authRoutes);

// Handle 404 errors
app.use(notFoundHandler);

// Global error handler
app.use(errorHandler);

// Initialize database connection
const initializeDatabase = async () => {
  try {
    await AppDataSource.initialize();
    logger.info('Database connected successfully');
  } catch (error) {
    logger.error('Database connection failed:', error);
    process.exit(1);
  }
};

// Graceful shutdown
process.on('SIGTERM', async () => {
  logger.info('SIGTERM received, shutting down gracefully');
  try {
    await AppDataSource.destroy();
    logger.info('Database connection closed');
    process.exit(0);
  } catch (error) {
    logger.error('Error during shutdown:', error);
    process.exit(1);
  }
});

export { app, initializeDatabase };