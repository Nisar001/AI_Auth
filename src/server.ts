import { app, initializeDatabase } from './app';
import logger from './utils/logger';

const PORT = parseInt(process.env.PORT || '3000', 10);

const startServer = async () => {
  try {
    // Initialize database connection
    await initializeDatabase();
    
    // Start the server
    const server = app.listen(PORT, () => {
      logger.info(`🚀 Server running on port ${PORT}`);
      logger.info(`📚 Environment: ${process.env.NODE_ENV || 'development'}`);
      logger.info(`🔗 Health check: http://localhost:${PORT}/health`);
      logger.info(`🔐 Auth API: http://localhost:${PORT}/api/auth`);
    });

    // Handle server errors
    server.on('error', (error: any) => {
      if (error.code === 'EADDRINUSE') {
        logger.error(`Port ${PORT} is already in use`);
      } else {
        logger.error('Server error:', error);
      }
      process.exit(1);
    });

    // Graceful shutdown on SIGINT
    process.on('SIGINT', () => {
      logger.info('SIGINT received, shutting down gracefully');
      server.close(() => {
        logger.info('Server closed');
        process.exit(0);
      });
    });

  } catch (error) {
    logger.error('Failed to start server:', error);
    process.exit(1);
  }
};

// Start the server
startServer();