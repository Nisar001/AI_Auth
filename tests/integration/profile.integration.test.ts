import request from 'supertest';
import { app } from '../../src/app';
import { AppDataSource } from '../../src/config/db';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.test' });

jest.mock('../../src/middlewares/auth', () => {
  return {
    authenticateToken: jest.fn((req, res, next) => {
      req.user = {
        id: 1,
        email: 'profileuser@example.com',
        profile: { name: 'Test User' },
      };
      next();
    }),
    requireEmailVerification: jest.fn((req, res, next) => {
      if (!req.user?.email) {
        res.status(403).json({ error: 'Email verification required' });
        return;
      }
      next();
    }),
    requirePhoneVerification: jest.fn((req, res, next) => {
      if (!req.user?.profile?.name) {
        res.status(403).json({ error: 'Phone verification required' });
        return;
      }
      next();
    }),
  };
});

jest.mock('../../src/config/db', () => {
  return {
    AppDataSource: {
      getRepository: jest.fn(() => {
        return {
          create: jest.fn((entity: any) => {
            return { id: 1, ...entity };
          }),
          save: jest.fn(async (entity: any) => {
            return { id: 1, ...entity };
          }),
          findOne: jest.fn(async (query: any) => {
            if (query.where.id === 1) {
              return {
                id: 1,
                fname: 'Test',
                lname: 'User',
                email: 'profileuser@example.com',
                profile: { name: 'Test User' },
              };
            }
            return null;
          }),
        };
      }),
      initialize: jest.fn().mockResolvedValue(undefined),
      synchronize: jest.fn().mockResolvedValue(undefined),
      destroy: jest.fn().mockResolvedValue(undefined),
    },
  };
});

jest.mock('../../src/services/smsService', () => {
  return {
    SmsService: jest.fn().mockImplementation(() => {
      return {
        sendSms: jest.fn(async () => {
          return { success: true, message: 'Mock SMS sent' };
        }),
      };
    }),
  };
});

jest.mock('../../src/services/emailService', () => {
  return {
    EmailService: jest.fn().mockImplementation(() => {
      return {
        sendEmail: jest.fn(async () => {
          return { success: true, message: 'Mock email sent' };
        }),
      };
    }),
  };
});

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Profile Integration', () => {
  let accessToken: string;

  beforeEach(() => {
    accessToken = 'mock-access-token';
  });

  it('should get user profile', async () => {
    const res = await request(app)
      .get('/api/auth/profile')
      .set('Authorization', `Bearer ${accessToken}`);

    expect([200, 201]).toContain(res.status);
    expect(res.body.success).toBe(true);
    expect(res.body.data.user.email).toBe('profileuser@example.com');
    expect(res.body).toMatchSnapshot();
  });

  it('should update user profile', async () => {
    const res = await request(app)
      .put('/api/auth/profile')
      .set('Authorization', `Bearer ${accessToken}`)
      .send({ fname: 'Updated', lname: 'Name' });

    expect([200, 201]).toContain(res.status);
    expect(res.body.success).toBe(true);
    expect(res.body.data.user.fname).toBe('Updated');
    expect(res.body).toMatchSnapshot();
  });

  // Refactor the profile retrieval test
  it('should retrieve user profile successfully', async () => {
    const repository = AppDataSource.getRepository('User');

    // Mock repository methods
    jest.spyOn(repository, 'create').mockImplementation((entity: any) => {
      return { id: 1, ...entity };
    });

    jest.spyOn(repository, 'save').mockImplementation(async (entity: any) => {
      return { id: 1, ...entity };
    });

    jest.spyOn(repository, 'findOne').mockImplementation(async (query: any) => {
      if (query.where.email === 'test@example.com') {
        return { id: 1, email: 'test@example.com', profile: { name: 'Test User' } };
      }
      return null;
    });

    // Test logic
    const user = repository.create({ email: 'test@example.com', password: 'password123', profile: { name: 'Test User' } });
    await repository.save(user);

    const savedUser = await repository.findOne({ where: { email: 'test@example.com' }, relations: ['profile'] });
    expect(savedUser).not.toBeNull();
    expect(savedUser!.profile.name).toBe('Test User');
    expect(savedUser).toMatchSnapshot({
      id: expect.any(Number),
      email: expect.any(String),
      profile: {
        name: expect.any(String),
      },
    });
  });
});
