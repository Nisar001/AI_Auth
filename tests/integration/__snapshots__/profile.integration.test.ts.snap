// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Integration should get user profile 1`] = `
{
  "data": {
    "user": {
      "address": {},
      "email": "profileuser@example.com",
      "fname": "Test",
      "id": 1,
      "lname": "User",
    },
  },
  "message": "Profile retrieved successfully",
  "statusCode": 200,
  "success": true,
}
`;

exports[`Profile Integration should retrieve user profile successfully 1`] = `
{
  "email": Any<String>,
  "id": Any<Number>,
  "profile": {
    "name": Any<String>,
  },
}
`;

exports[`Profile Integration should update user profile 1`] = `
{
  "data": {
    "user": {
      "address": {},
      "email": "profileuser@example.com",
      "fname": "Updated",
      "id": 1,
      "lname": "Name",
    },
  },
  "message": "Profile updated successfully",
  "statusCode": 200,
  "success": true,
}
`;
